### PR TITLE
compatible SharpRaven to .net 2.0

### PR DESCRIPTION
**Pull request by ArthurRyan0803**
1. Add genric deleagte : Func
2. Add TimeZomeInfo class , reference : https://www.codeproject.com/Articles/26727/How-to-use-time-zones-in-NET-under-Windows
3. Add extension method : Enumerable class & ExtensionAttribute

### CODE REVIEW

**Comment by bruno-garcia** [comment]
@ArthurRen There are quite a few failing tests though. Could you please get the builds fixed?
Other then that, LGTM.

Would be great to get some feedback from @asbjornu

**Comment by bruno-garcia** [comment]
@ArthurRen There are quite a few failing tests though. Could you please get the builds fixed?
Other then that, LGTM.

Would be great to get some feedback from @asbjornu

File: src/app/SharpRaven/Data/BackgroundRequester.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿#if !NET40 && !NET35
+﻿#if !NET40 && !NET35 && !NET20
```
Comment (asbjornu) [inline src/app/SharpRaven/Data/BackgroundRequester.cs:1]: `NET20` is ambigous. If we introduce a compile time flag for .NET Core, I would prefer if it was named `NETCORE20` or `NETSTANDARD20` and then we should probably rename all of the old flags to `NETFX` for consistency and to be explicit.

Comment (asbjornu) [inline src/app/SharpRaven/Data/BackgroundRequester.cs:1]: `NET20` is ambigous. Does it mean .NET Framework 2.0 or .NET Standard 2.0? If we introduce a compile time flag for .NET Core, I would prefer if it was named `NETCORE20` or `NETSTANDARD20` and then we should probably rename all of the old flags to `NETFX` for consistency and to be explicit.

File: src/app/SharpRaven/Data/HttpRequestBodyConverter.cs [hunk 2]
```diff
@@ -49,9 +49,9 @@ public static class HttpRequestBodyConverter
         /// A structured type for the specified <paramref name="httpContext"/>'s request body
         /// or <c>null</c> if the <paramref name="httpContext"/> is null, or the somehow conversion fails.
         /// </returns>
-        #if NET35
+#if NET35 || NET20
```
Comment (asbjornu) [inline src/app/SharpRaven/Data/HttpRequestBodyConverter.cs:52]: It seems like the consistent way to have compile time flags is to align them with the context and not with the left margin. I don't care where they are placed, I just want it to be consistent. :)

Comment (asbjornu) [inline src/app/SharpRaven/Data/HttpRequestBodyConverter.cs:52]: It seems like the consistent way to have compile time flags in the solution is to align them with the context and not with the left margin. I don't care where they are placed, I just want it to be consistent. :)

File: src/app/SharpRaven/SharpRaven.csproj [hunk 3]
```diff
@@ -1,8 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net471;net45;net40;net35</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;net471;net45;net40;net35;net20</TargetFrameworks>
     <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
-    <Configurations>Debug;Release;Debug-net35;Release-net35;Debug-net40;Release-net40;Debug-net45;Release-net45;Debug-net471;Release-net471;Release-netstandard2.0;Debug-netstandard2.0;Release-netstandard2.0</Configurations>
+    <Configurations>Debug;Release;Debug-net20;Release-net20;Debug-net35;Release-net35;Debug-net40;Release-net40;Debug-net45;Release-net45;Debug-net471;Release-net471;Release-netstandard2.0;Debug-netstandard2.0;Release-netstandard2.0</Configurations>
```
Comment (asbjornu) [inline src/app/SharpRaven/SharpRaven.csproj:5]: Same comment applies here; the name should be `netcore20` or `netstandard20`.

File: src/app/SharpRaven/Utilities/CircularBuffer.cs [hunk 4]
```diff
@@ -1,25 +1,27 @@
...
 using System.Linq;
+#endif
 
 namespace SharpRaven.Utilities {
     public class CircularBuffer<T>
     {
-        #if NET35
-        private class ConcurrentQueue<T> : ICollection, IEnumerable<T>
+#if NET35 || NET20
+        private class ConcurrentQueue<TElement> : ICollection, IEnumerable<TElement>
```
Comment (asbjornu) [inline src/app/SharpRaven/Utilities/CircularBuffer.cs:15]: Is this change from `T` to `TElement` necessary?

Comment (asbjornu) [inline src/app/SharpRaven/Utilities/CircularBuffer.cs:15]: Is this change from `T` to `TElement` necessary?

**Review by asbjornu — CHANGES_REQUESTED** [review]
Thanks for this!

I'm not sure this is the right direction to move the SharpRaven library in, though, since [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) is the forward looking library @getsentry is focusing on. That library already has .NET Core 2.0 support, doesn't it, @bruno-garcia? On top of that, [.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/), so 2.1 seems like a more forward-looking target.

Also, the tests are failing, so those need to be fixed before anything can be merged. I have a few comments on the naming and formatting, but otherwise this looks good.

**Review by asbjornu — COMMENTED** [review]
[.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/) and [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) already has .NET Core 2.0 support, so I'm not sure this pull request has much value, sorry.

**Comment by asbjornu** [comment]
[.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/) and [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) already has .NET Core 2.0 support, so I'm not sure this pull request has much value, sorry.

**Review by asbjornu — CHANGES_REQUESTED** [review]
[.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/) and [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) afaik already has .NET Core 2.0 support, so I'm not sure this PR is necessary?

For this to be merged, I would want the following to change:

- Target 2.1 instead of 2.0
- Fix my review comments on naming, ambiguity and consistency.
- Ensure the build is successful.

If that is taken care of, this would have 👍 from me. 😃

**Review by asbjornu — CHANGES_REQUESTED** [review]
Thanks for this!

I'm not sure this is the right direction to move the SharpRaven library in, though, since [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) is the forward looking library @getsentry is focusing on. That library already has .NET Core 2.0 support, doesn't it, @bruno-garcia? On top of that, [.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/), so 2.1 seems like a more forward-looking target.

Also, the tests are failing, so those need to be fixed before anything can be merged. I have a few comments on the naming and formatting, but otherwise this looks good.

**Review by asbjornu — CHANGES_REQUESTED** [review]
Thanks for this!

I'm not sure this is the right direction to move the SharpRaven library in, though, since [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) is the forward looking library @getsentry is focusing on. That library already has .NET Core 2.0 support, doesn't it, @bruno-garcia? On top of that, [.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/), so 2.1 seems like a more forward-looking target.

Also, the tests are failing, so those need to be fixed before anything can be merged. I have a few comments on the naming and formatting, but otherwise this looks good.

**Review by asbjornu — CHANGES_REQUESTED** [review]
Thanks for this!

I'm not sure this is the right direction to move the SharpRaven library in, though, since [sentry-dotnet](https://github.com/getsentry/sentry-dotnet) is the forward looking library @getsentry is focusing on. That library already has .NET Core 2.0 support, doesn't it, @bruno-garcia? On top of that, [.NET Core 2.0 reached end of life October 1st](https://blogs.msdn.microsoft.com/dotnet/2018/06/20/net-core-2-0-will-reach-end-of-life-on-september-1-2018/), so 2.1 seems like a more forward-looking target.

Also, the tests are failing, so those need to be fixed before anything can be merged. I have a few comments on the naming and formatting, but otherwise this looks good.

**Comment by ArthurRyan0803** [comment]
Sorry , i'm busy with work recently , i will check out these problems at
this weekend.

Bruno Garcia <notifications@github.com> 于2018年10月22日周一 下午4:06写道：

**Comment by bruno-garcia** [comment]
@asbjornu That is correct that the [unified API SDK](https://github.com/getsentry/sentry-dotnet) supports .NET Core 2.0 and also its EOL. Please note that this PR is supposed to add support to .NET Framework 2.0 (2005).

Since the jump from the already supported .NET Framework 3.5 to 2.0 doesn't seem very big (as shows in the code change here) and requests to support 2.0 came 2 or 3 times. 

Also this change doesn't affect anyone using any version of .NET newer than 2.0 since the artifacts generated are split per version, I'd be OK with merging this in if you are. That is, if the builds are fixed.

**Comment by asbjornu** [comment]
Aha, sorry. Yesterday was a bad day doing code reviews. I withdraw my comments minus the failing tests, then. If they are fixed, this LGTM.

**Pull request closed without merging**